src/run.rs

```diff
@@ -84,39 +84,62 @@ impl WorkerContext {
         let id = self.operation_counter.fetch_add(1, Ordering::Relaxed);
         (id < INVALID_OP_ID_THRESHOLD).then_some(id)
     }
+}
 
-    // Repeatedly runs the `operation` until it is asked to stop
-    // or an execution of the `operation` will either return `Err`
-    // or `ControlFlow::Break`.
-    pub async fn run_worker(&self, mut operation: Box<dyn Operation>) -> Result<()> {
-        'ops: while let Some(op_id) = self.issue_operation_id() {
-            'trying: for trial_idx in 0.. {
-                let scheduled_start_time = if let Some(rate_limiter) = &self.rate_limiter {
-                    let start_time = rate_limiter.issue_next_start_time();
-                    tokio::time::sleep_until(start_time).await;
-                    start_time
-                } else {
-                    Instant::now()
-                };
-                let actual_start_time = Instant::now();
-
-                let ctx = OperationContext {
-                    operation_id: op_id,
-                    scheduled_start_time,
-                    actual_start_time,
-                };
-
-                match operation.execute(&ctx).await {
-                    Ok(ControlFlow::Continue(_)) => continue 'ops,
-                    Ok(ControlFlow::Break(_)) => break 'ops,
-                    Err(err) if trial_idx >= self.max_retries_per_op => return Err(err),
-                    Err(err) if self.should_stop() => return Err(err),
-                    Err(_) => continue 'trying,
-                }
-            }
+pub struct WorkerSession {
+    context: Arc<WorkerContext>,
+    op_id: u64,
+    trial_idx: usize,
+}
+
+// Not the most beautiful interface, but it works - unlike async callbacks,
+// which I also tried, but failed to make the types work.
+impl WorkerSession {
+    fn new(context: Arc<WorkerContext>) -> Self {
+        Self {
+            context,
+            op_id: 0,
+            trial_idx: 0,
         }
+    }
+
+    // Should be called before starting an operation.
+    pub async fn start(&mut self) -> Option<OperationContext> {
```

Comment:
I got confuse by this methods name.
Maybe it could be renamed to `start_operation` and `end_operation`?
Otherwise it looks like we start a `WorkerSession` a million times.

Reply:
Good point, renamed to `start_operation` and `end_operation`.